src/main/default/classes/TriggerHandler.cls

```diff
@@ -65,51 +73,66 @@ public class TriggerHandler {
-        disabledObjects.clear();
-        disabledHandlers.clear();
-        disabledObjectHandlers.clear();
-        for (Trigger_Control__mdt triggerControl : triggerControls) {
-            String objectName = triggerControl.Object__c == null ? null : triggerControl.Object__c.toLowerCase();
-            String handlerName = triggerControl.Class__c == null ? null : triggerControl.Class__c.toLowerCase();
-            if (handlerName == null && objectName != null) {
-                disabledObjects.add(objectName);
-            } else if (handlerName != null && objectName == null) {
-                disabledHandlers.add(handlerName);
-            } else if (handlerName != null && objectName != null) {
-                if (!disabledObjectHandlers.containsKey(objectName)) {
-                    disabledObjectHandlers.put(objectName, new Set<String>());
-                }
-                disabledObjectHandlers.get(objectName).add(handlerName);
+    private void doAutomaticBinding() {
+        if (!triggerConfig.enabled) {
+            return;
+        }
+        for (String handlerName : triggerConfig.handlerOrder) {
+            TriggerConfigService.HandlerConfig handlerConfig = triggerConfig.handlers.get(handlerName);
+            // do not bind if auto bind is off or if handler is disabled
+            if (!handlerConfig.autoBind || !handlerConfig.enabled) {
+                continue;
+            }
+            Type t = getType(handlerConfig.className);
+            if (t == null) {
+                System.debug(handlerConfig.className + '? Looks like that handler class does not exist!');
+                continue;
+            }
+            HandlerInterface handlerClass = (HandlerInterface) t.newInstance();
+            if (handlerConfig.beforeInsert) {
+                bind(Evt.beforeInsert, handlerClass);
+            }
+            if (handlerConfig.afterInsert) {
+                bind(Evt.afterInsert, handlerClass);
+            }
+            if (handlerConfig.beforeUpdate) {
+                bind(Evt.beforeUpdate, handlerClass);
+            }
+            if (handlerConfig.afterUpdate) {
+                bind(Evt.afterUpdate, handlerClass);
+            }
+            if (handlerConfig.beforeDelete) {
+                bind(Evt.beforeDelete, handlerClass);
+            }
+            if (handlerConfig.afterDelete) {
+                bind(Evt.afterDelete, handlerClass);
+            }
+            if (handlerConfig.afterUndelete) {
+                bind(Evt.afterUndelete, handlerClass);
             }
         }
     }
-    static {
-        cacheTriggerControl([SELECT Class__c, Object__c FROM Trigger_Control__mdt]);
-    }
 
-    private Boolean objectDisabled(String xSObjType) {
-        return disabledObjects.contains(xSObjType);
+    private Type getType(String className) {
+        Type t;
+        try {
+            t = Type.forName(className); // doesn't throw exception if invalid name!
+        } catch (Exception e) {
```

Comment:
nitpick, don't catch generic exception here, catch the specific exception, and report a good error message as to what that exception means ... (opinions may vary here)

Reply:
personally i think it would be better to have one global try/catch in the externally visible methods (constructor/bind/manage) that catch anything we didn't think of before hand and report the details as "Unknown exception during {bind/manage/init}: {e}"

Reply:
^ you probably weren't implying this, but we obviously need to avoid wrapping calls to the `handle()` in a try/catch

Reply:
i wasn't implying that, but now that you mention i'd think we would want that. if one handler has an exception I'd rather have that get caught so the rest of the handlers could continue

side note, the error handling in this would be a great use case for platform events, that would let admins create flows to take action on trigger errors

side note2, could see the error behavior being a desirable config point (catch/throw, rollback/no rollback)

Reply:
Thinking this warrants a convo. As you can tell - I sort of paused on trying to handle errors.

I think there are two good options. Either:
- We do as little as possible and only handle errors for the logic we are managing in the auto binding and disabling. Fire off some sort of notification so admins can actually do something.
- We go bigger and actually build an trigger error object and add some error options like Ralph said above. On error, just log and continue? Throw error and rollback?

Reply:
we need this to not blow up on people first and foremost, anything that ends up preventing people from updating their opportunities is a major deal breaker

beyond that everything else sounds like a future need

Reply:
@ralphcallaway blindingly catching the `handle()` method of all trigger means that you would never have a trigger throw an exception.  While 99% of the time it's probably a better practice to use `addError`... blindly catching exceptions could cause serious data integrity issue.

That would be a massive functional departure from how it works today.

Reply:
Agree - save the additional error strategies for later.

But where do we put errors for typos or handlers without the handler interface implemented? I feel like debugging is not enough feedback, but also want to allow the rest of the valid handlers to run if we just have one bad piece of config. I just wouldn't want to let someone think that their handler is working and only know about the error if they go read the debug log.

Also don't want to try to stuff more functionality into this batch of stuff... maybe the right answer is actually throwing an error for bad config for now?

Reply:
I think for now, just system.debug out... Platform event's might be an option, but we need to do it in a way where we can be 100% sure our "debugging" won't break thing.

Reply:
@ChuckJonas the big difference in my mind is that triggers are now being altered without a deployment, so no test method changes, and potentially no integration testing. in any event, i disagree and commit to letting unhandled exceptions from the handler alone and having them bubble up

Reply:
For now just debugging duplicate bindings, invalid class names, and classes with no handler interface.

Allowing the handle() method to throw errors since if the trigger is actually hitting data errors and admin could simply use the metadata to turn the errant handler off! 😄